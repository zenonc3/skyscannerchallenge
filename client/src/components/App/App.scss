@import '~bpk-mixins';

.App {
  &__main {
    min-height: 100vh;
    padding: $bpk-spacing-sm;
    background-color: grayscale($color: #eee);
  }

  &__spinner {
    margin-top: 25vh;
    text-align: center;
  }
}
