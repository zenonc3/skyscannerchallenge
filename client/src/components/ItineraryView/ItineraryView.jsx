import PropTypes from 'prop-types';
import React from 'react';
import Itinerary from './../Itinerary';


const ItineraryView = ({ itineraries }) => {
  const itineraryCards = [];
  itineraries.forEach((itinerary) => {
    itineraryCards.push(<Itinerary key={itinerary.pricing.bookingUrl} {...itinerary} />);
  });
  return itineraryCards.length > 100 ? itineraryCards.slice(0, 100) : itineraryCards;
};

ItineraryView.propTypes = {
  itineraries: PropTypes.arrayOf(Itinerary).isRequired,
};

export default ItineraryView;
