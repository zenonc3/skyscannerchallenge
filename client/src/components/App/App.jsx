import BpkText from 'bpk-component-text/src/BpkText';
import { BpkExtraLargeSpinner } from 'bpk-component-spinner';
import React from 'react';
import STYLES from './App.scss';
import Header from './../Header';
import ItineraryView from '../ItineraryView';
import SearchSummary from '../SearchSummary';
import ActionBar from '../ActionBar/ActionBar';
import { getNextMonday, getNextDay } from './../../utils/utils';

const c = className => STYLES[className] || 'UNKNOWN';

class App extends React.Component {
  constructor() {
    super();
    this.query = {
      origin: 'LHR',
      destination: 'EDI',
      numTravellers: 2,
      outboundDate: getNextMonday(),
      inboundDate: getNextDay(getNextMonday()),
      cabinClass: 'Economy',
    };
    this.state = {
      itineraries: [],
      isLoaded: false,
    };
  }

  componentDidMount() {
    const url = new URL('http://localhost:4000/api/search');
    url.searchParams.append('origin', this.query.origin);
    url.searchParams.append('destination', this.query.destination);
    url.searchParams.append('adults', this.query.numTravellers);
    url.searchParams.append('outboundDate', this.query.outboundDate);
    url.searchParams.append('inboundDate', this.query.outboundDate);
    url.searchParams.append('cabinClass', this.query.cabinClass);
    fetch(url)
      .then(response => response.json())
      .then((results) => {
        this.setState({
          itineraries: results.itineraries || [],
          isLoaded: true,
        });
      })
      .catch(console.error);
  }

  render() {
    const { itineraries, isLoaded } = this.state;
    return (
      <div className={c('App')}>
        <Header />
        <SearchSummary
          origin={this.query.origin}
          destination={this.query.destination}
          numTravellers={this.query.numTravellers}
          cabinClass={this.query.cabinClass}
        />
        <ActionBar />
        <main className={c('App__main')}>
          {isLoaded ? <ItineraryView itineraries={itineraries} />
                    : <div className={c('App__spinner')}><BpkExtraLargeSpinner /> <BpkText tagName="p">This could take a few seconds...</BpkText> </div> }
        </main>
      </div>
    );
  }
}

export default App;
