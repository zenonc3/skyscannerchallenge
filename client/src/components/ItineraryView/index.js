import ItineraryView from './ItineraryView';

export default ItineraryView;
