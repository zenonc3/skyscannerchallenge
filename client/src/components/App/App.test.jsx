import React from 'react';
import { BpkExtraLargeSpinner } from 'bpk-component-spinner';
import ShallowRenderer from 'react-test-renderer/shallow';
import { create } from 'react-test-renderer';

import App from './App';
import ItineraryView from '../ItineraryView';

describe('App', () => {
  beforeAll(() => {
    global.URL = jest.fn(() => ({
      searchParams: {
        append: jest.fn(),
      },
    }));

    const mockSuccessResponse = {};
    const mockJsonPromise = Promise.resolve(mockSuccessResponse);
    const mockFetchPromise = Promise.resolve({
      json: () => mockJsonPromise,
    });
    jest.spyOn(global, 'fetch').mockImplementation(() => mockFetchPromise);
  });

  it('should render correctly', () => {
    const renderer = new ShallowRenderer();
    renderer.render(<App />);
    expect(renderer.getRenderOutput()).toMatchSnapshot();
  });

  it('should render a spinner before the data is loaded', () => {
    const { root } = create(<App />);
    expect(root.instance.state.isLoaded).toBeFalsy();
    expect(root.findByType(BpkExtraLargeSpinner)).toBeTruthy();
  });

  it('should render an Itinerary View after the data is loaded', () => {
    const { root } = create(<App />);
    root.instance.setState({
      isLoaded: true,
      itineraries: [],
    });
    expect(root.findByType(ItineraryView)).toBeTruthy();
  });
});

