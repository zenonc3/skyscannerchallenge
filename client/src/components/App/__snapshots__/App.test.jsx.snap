// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should render correctly 1`] = `
<div
  className="App"
>
  <Header />
  <SearchSummary
    cabinClass="Economy"
    destination="EDI"
    numTravellers={2}
    origin="LHR"
  />
  <ActionBar />
  <main
    className="App__main"
  >
    <div
      className="App__spinner"
    >
      <BpkExtraLargeSpinner
        className={null}
        type="dark"
      />
       
      <BpkText
        bold={false}
        className={null}
        tagName="p"
        textStyle="base"
      >
        This could take a few seconds...
      </BpkText>
       
    </div>
  </main>
</div>
`;
